File: exercises/cpp/lab06/dining_philosopher.cpp
// Lab 06 - Dining Philosopher Problem using Semaphores
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define N         5
#define THINKING  0
#define HUNGRY    1
#define EATING    2
#define LEFT      ((id + N - 1) % N)
#define RIGHT     ((id + 1) % N)

int state[N];
int phil[N] = {0, 1, 2, 3, 4};

sem_t mutex;
sem_t S[N];

void test(int id) {
    if (state[id] == HUNGRY &&
        state[LEFT]  != EATING &&
        state[RIGHT] != EATING) {
        state[id] = EATING;
        printf("Philosopher %d takes fork %d and %d\n", id + 1, LEFT + 1, id + 1);
        printf("Philosopher %d is Eating\n", id + 1);
        sem_post(&S[id]);
    }
}

void take_fork(int id) {
    sem_wait(&mutex);
    state[id] = HUNGRY;
    printf("Philosopher %d is Hungry\n", id + 1);
    test(id);
    sem_post(&mutex);
    sem_wait(&S[id]);
    sleep(1);
}

void put_fork(int id) {
    sem_wait(&mutex);
    state[id] = THINKING;
    printf("Philosopher %d putting fork %d and %d down\n", id + 1, LEFT + 1, id + 1);
    printf("Philosopher %d is Thinking\n", id + 1);
    test(LEFT);
    test(RIGHT);
    sem_post(&mutex);
}

void* philosopher(void* arg) {
    int id = *(int*)arg;
    // Each philosopher eats 3 times then stops (avoid infinite loop in demo)
    for (int i = 0; i < 3; i++) {
        sleep(1);
        take_fork(id);
        sleep(1);
        put_fork(id);
    }
    return NULL;
}

int main() {
    pthread_t thread_id[N];
    sem_init(&mutex, 0, 1);
    for (int i = 0; i < N; i++) sem_init(&S[i], 0, 0);

    printf("=== Dining Philosopher Problem (Semaphore) ===\n\n");
    for (int i = 0; i < N; i++) {
        state[i] = THINKING;
        printf("Philosopher %d is thinking\n", i + 1);
        pthread_create(&thread_id[i], NULL, philosopher, &phil[i]);
    }

    for (int i = 0; i < N; i++) pthread_join(thread_id[i], NULL);

    printf("\nAll philosophers finished.\n");
    sem_destroy(&mutex);
    for (int i = 0; i < N; i++) sem_destroy(&S[i]);
    return 0;
}

File: exercises/python/lab06/dining_philosopher.py
"""Lab 06 - Dining Philosopher Problem using semaphores"""
import threading
import time

N = 5
THINKING, HUNGRY, EATING = 0, 1, 2

state  = [THINKING] * N
mutex  = threading.Semaphore(1)
forks  = [threading.Semaphore(0) for _ in range(N)]


def left(i):  return (i + N - 1) % N
def right(i): return (i + 1) % N


def test(i):
    if state[i] == HUNGRY and state[left(i)] != EATING and state[right(i)] != EATING:
        state[i] = EATING
        print(f"Philosopher {i+1} takes fork {left(i)+1} and {i+1}")
        print(f"Philosopher {i+1} is Eating")
        forks[i].release()


def take_fork(i):
    mutex.acquire()
    state[i] = HUNGRY
    print(f"Philosopher {i+1} is Hungry")
    test(i)
    mutex.release()
    forks[i].acquire()
    time.sleep(0.5)


def put_fork(i):
    mutex.acquire()
    state[i] = THINKING
    print(f"Philosopher {i+1} putting fork {left(i)+1} and {i+1} down")
    print(f"Philosopher {i+1} is Thinking")
    test(left(i))
    test(right(i))
    mutex.release()


def philosopher(i):
    for _ in range(3):   # each philosopher eats 3 times
        time.sleep(0.5)
        take_fork(i)
        time.sleep(0.5)
        put_fork(i)


if __name__ == "__main__":
    print("=== Dining Philosopher Problem (Semaphore) ===\n")
    threads = [threading.Thread(target=philosopher, args=(i,)) for i in range(N)]
    for i, t in enumerate(threads):
        print(f"Philosopher {i+1} is thinking")
        t.start()
    for t in threads: t.join()
    print("\nAll philosophers finished.")
